### ISSUE TITLE
Cannot reproduce test results

### ISSUE THREAD

**Issue by jbrestan** [fscheck/FsCheck#395]
I'm using FsCheck 2.9 and FsCheck.Xunit 2.9 to test a monadic wrapper around `Result` type. The complete code is below for inspection. In the last year or so, I've seen the test below fail twice (out of thousands of runs on team's dev machines and build servers) and only once we were able to save the error message containing the seed for value generation. If I use the seed to replay the failing test, the test replay always passes.

This is the full code as used in our codebase:

```
(* the custom Result type, just a regular DU, not a struct *)
type Result<'result, 'message> =
    | Ok of 'result
    | Error of 'message

(* bind implementation *)
let inline bind (f: 'a -> Result<'b, 'c>) (result: Result<'a, 'c>) =
    match result with
    | Ok x -> f x
    | Error e -> Error e

(* bind infix *)
let inline (>>=) r f = bind f r

(* test aliases since FsCheck doesn't support open generic types *)
type OkType = int
type ErrorType = string

(* the test itself *)
[<Property>]
let associativity
    (f: OkType -> Result<OkType, ErrorType>)
    (g: OkType -> Result<OkType, ErrorType>)
    (result: Result<OkType, ErrorType>) =

    (result >>= f) >>= g = (result >>= (fun v -> (f v) >>= g))
```

As I said, this test works literally 99.9% of times. The one case where I was able to obtain the full error message is this:

> Result Message:               
> FsCheck.Xunit.PropertyFailedException : 
> Falsifiable, after 47 tests (0 shrinks) (StdGen (242196637,296351952)):
> Original:
> (<fun:Invoke@3253>, <fun:Invoke@3253>, Ok -12)

But when I tried to run the test manually with the specified StdGen, I wasn't able to reproduce the failed result. This is what I tried
```
Check.One({ Config.QuickThrowOnFailure with Replay = Some <| Random.StdGen (242196637,296351952) }, associativity)
```
I also tried moving the `MaxTest` and `EndSize` values from the default of 100 to 10000 and running it few dozen times, but still nothing.

Weirdly enough, there is another property that failed once, also using `bind`, but it's part of the result builder tests, so it would be a bit more code to provide here...

Am I doing something wrong when replaying the check?
What could possibly be failing there?
Is there something wrong with how I'm defining the property in the first place?
Or is it a bug in FsCheck's value generation somehow?

**Comment by kurtschelfthout** [comment]
Are you running tests in parallel? If you are generating functions, it uses this under the hood https://github.com/fscheck/FsCheck/blob/master/src/FsCheck/Gen.fs#L607 which is not thread safe.

That's about the only thing I can think of right now...

**Comment by kurtschelfthout** [comment]
By the way, you can also use the `Function` wrapper so you get more meaningful output than `fun:Invoked@3253`: https://fscheck.github.io/FsCheck/TipsAndTricks.html#Properties-of-functions

**Comment by kurtschelfthout** [comment]
I've released 2.10.1 which contains a thread-safe function generator, cherry-picked from fscheck3 branch.

Feel free to re-open if that doesn't help.

**Comment by jbrestan** [comment]
Thank you!
The original test failure happens on a colleague's machine, I wasn't able to find out the exact configuration. We'll try the updated version and see if it happens again.

**Comment by jbrestan** [comment]
Hmm, this may be a separate issue, possibly in xUnit repository instead of here, but updating to 2.10.1 makes xUnit's console runner (used by ReSharper, FAKE...) consistently fail with `System.AppDomainUnloadedException` on all FsCheck Properties. Have you ever encountered that?

I found similar issue in xunit repo, but it was related to a different runner, and seem to be resolved some time ago... https://github.com/xunit/xunit/issues/490

**Comment by kurtschelfthout** [comment]
No. FsCheck itself uses xUnit and it seems to work fine. Can you try to attach a debugger and get the actual error?

**Comment by kurtschelfthout** [comment]
Oh, perhaps it's the stackoverflow I fixed in 2.10.2 (and introduced in 2.10.0). Can you try 2.10.2 freshly released just now?

**Comment by jbrestan** [comment]
Still the same behavior with 2.10.2, I'll look into the xunit console runner when I have more time, hopefully by the end of the week

**Comment by jbrestan** [comment]
I've updated to 2.10.3, this time also upgraded FsCheck.Xunit, which I use, but forgot to update/assumed compatibility... and it works again now.

Unfortunately I don't have time to dig into the root cause here, but I suspect some of the project file cleanups or transitive dependency changes may have caused binary incompatibility between 2.9 and newer versions of FsCheck and FsCheck.Xunit.

I think it may be worth adding package version dependencies on the FsCheck modules like Xunit integration so that this upgrade happens automatically.

Thank you again for your help! I'll let you know if we see the same original issue on 2.10.2+, but I believe it to be solved now.

**Comment by kurtschelfthout** [comment]
Oh, right, thanks. I think FsCheck dep should basically be pinned to = instead of >=. Not sure where the incompatibility comes from though in 2.10 - I updated NUnit, but didn't touch anything in FsCheck.Xunit.

Anyway don't see any particular advantage to keeping the >= either, so will do that.
